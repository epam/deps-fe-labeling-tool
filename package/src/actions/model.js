
import { batch } from 'react-redux'
import { createAction } from 'redux-actions'
import { LabelType, LABEL_TYPE_NAME } from '@/models/Label'
import { Markup } from '@/models/Markup'
import { markupSelector } from '@/selectors/markup'
import { fieldsSelector, fieldsToDeleteSelector } from '@/selectors/model'
import {
  updateLabelsWithSettings,
  updateTablesWithSettings
} from './markup'

const FEATURE_NAME = 'MODEL'

const storeFields = createAction(
  `${FEATURE_NAME}/STORE`
)

const storeInitialFields = createAction(
  `${FEATURE_NAME}/STORE_INITIAL_FIELDS`
)

const updateInitialFields = createAction(
  `${FEATURE_NAME}/UPDATE_INITIAL_FIELDS`
)

const enableDeleteFieldsMode = createAction(
  `${FEATURE_NAME}/ENABLE_DELETE_FIELD_MODE`
)

const resetDefault = createAction(
  `${FEATURE_NAME}/RESET_DEFAULT`
)

const updateFieldsToDelete = createAction(
  `${FEATURE_NAME}/UPDATE_FIELDS_TO_DELETE`
)

const cancelDeleteFieldsMode = createAction(
  `${FEATURE_NAME}/CANCEL_DELETE_FIELDS_MODE`
)

const confirmFieldsDeletion = () => (dispatch, getState) => {
  const state = getState()
  const markup = markupSelector(state)
  const fields = fieldsSelector(state)
  const fieldsToDelete = fieldsToDeleteSelector(state)
  const newFields = fields.filter((field) => fieldsToDelete.every((f) => f.code !== field.code))
  const assignedObjects = markup
    ? Markup.getAllObjects(markup).filter((markupObject) => fieldsToDelete.find((field) => field.code === markupObject.fieldCode))
    : []

  batch(() => {
    assignedObjects.forEach((markupObject) => {
      if (markupObject.typeName === LABEL_TYPE_NAME) {
        dispatch(updateLabelsWithSettings({
          ...markupObject,
          fieldCode: '',
          type: LabelType.UNASSIGNED,
          index: undefined
        }))
      } else {
        dispatch(updateTablesWithSettings({
          ...markupObject,
          fieldCode: '',
          index: undefined
        }))
      }
    })
    dispatch(storeFields(newFields))
    dispatch(cancelDeleteFieldsMode())
  })
}

export {
  enableDeleteFieldsMode,
  resetDefault,
  storeFields,
  storeInitialFields,
  updateFieldsToDelete,
  updateInitialFields,
  confirmFieldsDeletion,
  cancelDeleteFieldsMode
}
