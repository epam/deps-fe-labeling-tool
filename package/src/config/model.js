import { FieldType } from '@/enums/FieldType'
import { Field } from '@/models/Field'
import { PairFieldMeta, ListFieldMeta, EnumFieldMeta } from '@/models/FieldMeta'

const fields = [
  new Field(
    'dates',
    'Dates',
    FieldType.DATE
  ),
  new Field(
    'firstName',
    'First Name',
    FieldType.STRING
  ),
  new Field(
    'lastName',
    'Last Name',
    FieldType.STRING,
    undefined,
    true
  ),
  new Field(
    'companyName',
    'Company Name',
    FieldType.STRING,
    undefined,
    true
  ),
  new Field(
    'city',
    'City',
    FieldType.ENUM,
    new EnumFieldMeta([
      'Anytown',
      'Sometown',
      'San Diego',
      'Los Angeles'
    ]),
    true
  ),
  new Field(
    'countryStates',
    'Country States',
    FieldType.LIST,
    new ListFieldMeta(FieldType.STRING),
    true
  ),
  new Field(
    'dateMatching',
    'Date Matching',
    FieldType.LIST,
    new ListFieldMeta(FieldType.CHECKMARK),
    true
  ),
  new Field(
    'date',
    'Date',
    FieldType.PAIR,
    new PairFieldMeta(FieldType.STRING, FieldType.STRING)
  ),
  new Field(
    'zipCodes',
    'Zip Codes',
    FieldType.LIST,
    new ListFieldMeta(
      FieldType.PAIR,
      new PairFieldMeta(FieldType.STRING, FieldType.STRING)
    ),
    true
  ),
  new Field(
    'homeAddress',
    'Home Address',
    FieldType.STRING
  ),
  new Field(
    'businessAddress',
    'Business Address',
    FieldType.PAIR,
    new PairFieldMeta(FieldType.STRING, FieldType.STRING),
    true
  ),
  new Field(
    'taxpayerSpousesInfo',
    'Taxpayer Spouse\'s Info',
    FieldType.TABLE
  ),
  new Field(
    'taxpayerInfo',
    'Taxpayer Info',
    FieldType.TABLE,
    undefined,
    true
  ),
  new Field(
    'tables',
    'Tables',
    FieldType.LIST,
    new ListFieldMeta(FieldType.TABLE)
  ),
  new Field(
    'uniqueIdentifier',
    'Unique Identifier',
    FieldType.PAIR,
    new PairFieldMeta(FieldType.STRING, FieldType.STRING)
  ),
  new Field(
    'signatory',
    'Signatory',
    FieldType.CHECKMARK
  ),
  new Field(
    'returnTranscript',
    'Return Transcript',
    FieldType.CHECKMARK
  ),
  new Field(
    'formNumbers',
    'Form Numbers',
    FieldType.LIST,
    new ListFieldMeta(
      FieldType.ENUM,
      new EnumFieldMeta([
        '1098',
        '5498',
        '1040'
      ])
    )
  )
]

export {
  fields
}
