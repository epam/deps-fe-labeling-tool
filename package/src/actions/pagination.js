import { createAction } from 'redux-actions'

const FEATURE_NAME = 'PAGINATION'

const openPage = createAction(`${FEATURE_NAME}/OPEN_PAGE`)

const resetDefault = createAction(
  `${FEATURE_NAME}/RESET_DEFAULT`
)

export {
  resetDefault,
  openPage
}
