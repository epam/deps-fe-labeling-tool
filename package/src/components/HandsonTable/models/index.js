
export * from './HTTableData'
export * from './HTMerge'
export * from './HTColumn'
