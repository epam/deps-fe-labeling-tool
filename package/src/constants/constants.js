const CONFIDENCE_ON_MANUAL_CHANGE = 1

export {
  CONFIDENCE_ON_MANUAL_CHANGE
}
