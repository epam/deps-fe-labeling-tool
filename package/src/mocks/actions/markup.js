const addLabels = jest.fn((labels) => ({
  type: 'MOCK_ADD_LABELS_ACTION',
  payload: labels
}))

const addTables = jest.fn((tables) => ({
  type: 'MOCK_ADD_TABLES_ACTION',
  payload: tables
}))

const removeLabels = jest.fn((labels) => ({
  type: 'MOCK_REMOVE_LABELS_ACTION',
  payload: labels
}))

const updateLabels = jest.fn((labels) => ({
  type: 'MOCK_UPDATE_LABELS_ACTION',
  payload: labels
}))

const updateAllLabels = jest.fn((labels) => ({
  type: 'MOCK_UPDATE_ALL_LABELS_ACTION',
  payload: labels
}))

const updateTables = jest.fn((tables) => ({
  type: 'MOCK_UPDATE_TABLES_ACTION',
  payload: tables
}))

const updateAllTables = jest.fn((tables) => ({
  type: 'MOCK_UPDATE_ALL_TABLES_ACTION',
  payload: tables
}))

const selectLabels = jest.fn((labels) => ({
  type: 'MOCK_SELECT_LABELS_ACTION',
  payload: labels
}))

const selectTables = jest.fn((tables) => ({
  type: 'MOCK_SELECT_TABLES_ACTION',
  payload: tables
}))

const clearSelection = jest.fn(() => ({
  type: 'MOCK_CLEAR_SELECTION_ACTION'
}))

const removeTables = jest.fn((tables) => ({
  type: 'MOCK_REMOVE_TABLES_ACTION',
  payload: tables
}))

const importMarkup = jest.fn(() => ({
  type: 'MOCK_IMPORT_MARKUP'
}))

const exportMarkup = jest.fn(() => ({
  type: 'MOCK_EXPORT_MARKUP'
}))

const undo = jest.fn(() => ({
  type: 'MOCK_UNDO_ACTION'
}))

const redo = jest.fn(() => ({
  type: 'MOCK_REDO_ACTION'
}))

const updateInitialMarkup = jest.fn(() => ({
  type: 'MOCK_UPDATE_INITIAL_MARKUP_ACTION'
}))

const storeMarkup = jest.fn((markup) => ({
  type: 'MOCK_STORE_MARKUP_ACTION',
  payload: markup
}))

const storeAssignedMarkup = jest.fn((markup) => ({
  type: 'MOCK_UPDATE_MARKUP_ACTION',
  payload: markup
}))

const updatePageMarkup = jest.fn(() => ({
  type: 'UPDATE_PAGE_MARKUP'
}))

const reset = jest.fn(() => ({
  type: 'RESET_MARKUP_ACTION'
}))

const copyMarkup = jest.fn(() => ({
  type: 'COPY_MARKUP_ACTION'
}))

const pasteMarkup = jest.fn(() => ({
  type: 'PASTE_MARKUP_ACTION'
}))

const updateLabelsWithSettings = jest.fn(() => ({
  type: 'UPDATE_LABELS_WITH_SETTINGS'
}))

const updateTablesWithSettings = jest.fn(() => ({
  type: 'UPDATE_TABLES_WITH_SETTINGS'
}))

const mockMarkupActions = {
  addLabels,
  addTables,
  copyMarkup,
  pasteMarkup,
  removeLabels,
  reset,
  updateLabels,
  updateAllLabels,
  updateTables,
  updateAllTables,
  selectLabels,
  selectTables,
  clearSelection,
  removeTables,
  importMarkup,
  exportMarkup,
  undo,
  updateInitialMarkup,
  storeMarkup,
  storeAssignedMarkup,
  updatePageMarkup,
  redo,
  updateLabelsWithSettings,
  updateTablesWithSettings
}

export {
  mockMarkupActions
}
