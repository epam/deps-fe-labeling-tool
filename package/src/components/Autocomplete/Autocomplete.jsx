
import React, { useState, useMemo, useCallback, useEffect } from 'react'
import { AutoComplete as AntdAutocomplete } from 'antd'
import 'antd/lib/auto-complete/style/index.less'
import lodashDebounce from 'lodash/debounce'
import PropTypes from 'prop-types'
import { optionShape } from '@/models/Option'

const DEBOUNCE_TIME = 250

const RESOURCE_NOT_FOUND_CONTENT = 'Not found any matches'

const Autocomplete = ({
  options,
  onChange,
  notFoundContent,
  value,
  onFocus,
  width,
  disabled,
  allowClear
}) => {
  const [optionsForRender, setOptions] = useState(options)
  const [autocompleteValue, setAutocompleteValue] = useState()

  useEffect(() => {
    setAutocompleteValue(value)
  }, [value])

  const onSearch = (searchText) => {
    if (!searchText) {
      setOptions(options)
      return
    }

    setOptions(options.filter((o) => o.value.toLowerCase().includes(searchText.toLowerCase())))
  }

  const debouncedOnChange = useMemo(
    () => lodashDebounce(
      (val) => onChange(val),
      DEBOUNCE_TIME
    ),
    [onChange]
  )

  const onChangeAutocomplete = useCallback(
    (value) => {
      setAutocompleteValue(value)
      debouncedOnChange(value)
    },
    [debouncedOnChange]
  )

  return (
    <AntdAutocomplete
      allowClear={allowClear}
      defaultValue={value}
      disabled={disabled}
      notFoundContent={notFoundContent}
      onChange={onChangeAutocomplete}
      onFocus={onFocus}
      onSearch={onSearch}
      options={optionsForRender}
      style={
        {
          width: width
        }
      }
      value={autocompleteValue}
    />
  )
}

Autocomplete.defaultProps = {
  notFoundContent: RESOURCE_NOT_FOUND_CONTENT,
  width: '100%'
}

Autocomplete.propTypes = {
  notFoundContent: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  onFocus: PropTypes.func,
  options: PropTypes.arrayOf(optionShape).isRequired,
  value: PropTypes.string,
  width: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string
  ]),
  disabled: PropTypes.bool,
  allowClear: PropTypes.bool
}

export {
  Autocomplete
}
