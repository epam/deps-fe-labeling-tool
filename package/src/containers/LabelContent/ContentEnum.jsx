import React, { useCallback, useMemo } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { updateLabels } from '@/actions/markup'
import { Autocomplete } from '@/components/Autocomplete'
import { CONFIDENCE_ON_MANUAL_CHANGE } from '@/constants/constants'
import { FieldType } from '@/enums/FieldType'
import { fieldShape } from '@/models/Field'
import { labelShape } from '@/models/Label'
import { Option } from '@/models/Option'
import { fieldsSelector } from '@/selectors/model'
import { currentPageSelector } from '@/selectors/pagination'

const ContentEnum = ({
  currentPage,
  label,
  updateLabels,
  fields
}) => {
  const onContentChange = useCallback(
    (value) => {
      updateLabels(currentPage, [{
        ...label,
        content: value,
        confidence: CONFIDENCE_ON_MANUAL_CHANGE
      }])
    },
    [
      currentPage,
      label,
      updateLabels
    ])

  const enumOptions = useMemo(() => {
    const { fieldType, fieldMeta } = fields.find((field) => field.code === label.fieldCode)
    const { options } = fieldType === FieldType.LIST ? fieldMeta.baseTypeMeta : fieldMeta

    return options.map((value) => new Option(value))
  }, [fields, label.fieldCode])

  return (
    <Autocomplete
      onChange={onContentChange}
      value={label.content}
      options={enumOptions}
    />
  )
}

ContentEnum.propTypes = {
  currentPage: PropTypes.number.isRequired,
  label: labelShape.isRequired,
  updateLabels: PropTypes.func.isRequired,
  fields: PropTypes.arrayOf(fieldShape.isRequired)
}

const mapStateToProps = (state) => ({
  currentPage: currentPageSelector(state),
  fields: fieldsSelector(state)
})

const mapDispatchToProps = {
  updateLabels
}

const ConnectedComponent = connect(mapStateToProps, mapDispatchToProps)(ContentEnum)

export {
  ConnectedComponent as ContentEnum
}
