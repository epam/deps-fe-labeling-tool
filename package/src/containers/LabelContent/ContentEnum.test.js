import { mockMarkupActions } from '@/mocks/actions/markup'
import { mockReactRedux } from '@/mocks/mockReactRedux'
import { mockModelSelectors } from '@/mocks/selectors/model'
import { mockPaginationSelectors } from '@/mocks/selectors/pagination'
import React from 'react'
import { shallow } from 'enzyme'
import { updateLabels } from '@/actions/markup'
import { Autocomplete } from '@/components/Autocomplete'
import { CONFIDENCE_ON_MANUAL_CHANGE } from '@/constants/constants'
import { Label, LabelType } from '@/models/Label'
import { Option } from '@/models/Option'
import { ContentEnum } from './ContentEnum'

jest.mock('react-redux', () => mockReactRedux)
jest.mock('@/actions/markup', () => mockMarkupActions)
jest.mock('@/selectors/pagination', () => mockPaginationSelectors)
jest.mock('@/selectors/model', () => mockModelSelectors)

const {
  WrappedComponent,
  mapStateToProps,
  mapDispatchToProps
} = ContentEnum

describe('Container: ContentEnum', () => {
  describe('mapStateToProps', () => {
    it('should pass expected state as props to WrappedComponent', () => {
      const { props } = mapStateToProps()

      const expectedState = {
        currentPage: mockPaginationSelectors.currentPageSelector(),
        fields: mockModelSelectors.fieldsSelector()
      }

      expect(props).toEqual(expectedState)
    })
  })

  describe('mapDispatchToProps', () => {
    const { props } = mapDispatchToProps()

    it('should pass updateLabels action as prop to WrappedComponent', () => {
      props.updateLabels([])
      expect(updateLabels).nthCalledWith(1, [])
    })
  })

  let wrapper, defaultProps

  beforeEach(() => {
    defaultProps = {
      ...mapStateToProps().props,
      ...mapDispatchToProps().props,
      label: new Label(
        0.1,
        0.2,
        0.3,
        0.4,
        'enum',
        undefined,
        LabelType.ENUM,
        'test',
        { data: 'some data' },
        0.7
      )
    }

    wrapper = shallow(<WrappedComponent {...defaultProps} />)
  })

  it('should render correct layout based on props', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should call props updateLabels with correct argument in case of calling to Autocomplete onChange', () => {
    const radioGroupProps = wrapper.find(Autocomplete).props()
    radioGroupProps.onChange('newValue')
    expect(defaultProps.updateLabels).nthCalledWith(
      1,
      defaultProps.currentPage,
      [
        {
          ...defaultProps.label,
          content: 'newValue',
          confidence: CONFIDENCE_ON_MANUAL_CHANGE
        }
      ])
  })

  it('should set correct options to Autocomplete if there are list of enums', () => {
    const enumListFieldCode = 'enumList'
    const enumListField = defaultProps.fields.find((f) => f.code === enumListFieldCode)
    const options = enumListField.fieldMeta.baseTypeMeta.options.map((o) => new Option(o))

    defaultProps.label = new Label(
      0.1,
      0.2,
      0.3,
      0.4,
      enumListFieldCode,
      1,
      LabelType.ENUM,
      'test',
      { data: 'some data' },
      0.7
    )

    wrapper = shallow(<WrappedComponent {...defaultProps} />)

    expect(wrapper.find(Autocomplete).props().options).toEqual(options)
  })
})
