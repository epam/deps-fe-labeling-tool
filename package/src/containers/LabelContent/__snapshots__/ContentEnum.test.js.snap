// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container: ContentEnum should render correct layout based on props 1`] = `
<Autocomplete
  notFoundContent="Not found any matches"
  onChange={[Function]}
  options={
    Array [
      Option {
        "disabled": undefined,
        "text": "test",
        "value": "test",
      },
      Option {
        "disabled": undefined,
        "text": "test1",
        "value": "test1",
      },
    ]
  }
  value="test"
  width="100%"
/>
`;
